Cythonizing mcsolve

Ericgig opened the pull request:
This contain a new solver for `mcsolve`, `dopri5` which don't go through scipy and is faster for small system.

It also contain a new class for time-dependent Qobj. It allow to create an object from the usual list of Qobj and string/function. The object can then be called to obtain the Qobj at a time. It is used in `mcsolve`.

Vutshi commented on qutip/cy/inter.pyx:
```diff
@@ -0,0 +1,63 @@
+    cdef double a = (t/dt - ii)
+    cdef complex approx = 0.
+
+    approx += (-a**3 + 3*a**2 - 2*a) * 0.1666666666666666 * str_array_0[ii-1]
```
This is a bit inefficient for two reasons. First, calculation of powers (like  a**2) is substantially slower than simple multiplication (a*a). Second, it is inefficient and numerically unstable to calculate polynomials in such a direct way. One can use Horner form of a polynomial instead.

nonhermitian commented on qutip/cy/td_qobj_cy.pyx:
```diff
@@ -0,0 +1,583 @@
+cdef class cy_td_qobj(cy_qobj):
+    def __init__(self):
+        self.N_ops = 0
+        self.ops = <CSR_Matrix**> malloc(0 * sizeof(CSR_Matrix*))
```
Calls to `malloc` should be replaced with `PyDataMem_NEW(size_t size)` from `numpy/arrayobject.h`. This is because the `malloc` called by the complied Cython code may not be the same as the `malloc` used when building Python.  For example, using GCC to compile QuTiP on Win.  Freeing data should also be done with ` PyDataMem_FREE(void * ptr)`.  Learned this the hard way.

Ericgig replied:
I will make the replacement this week.

nonhermitian commented on qutip/td_qobj.py:
```diff
@@ -0,0 +1,1111 @@
+
+    *Examples*
+
+        H = [H0, [H1, 'sin(w*t)'], [H2, f1_t], [H3, np.sin(3*w*tlist)]]
```
As it stands now, we do not support mixing str and Python func formats.  Of course we could, but then all the solvers would need to add this support.

Ericgig replied:
While the different format work together, there are not many advantages to mix then. In some methods, there are optimizations for only func format or only str format.  Should I just remove it from the documentation?

nonhermitian commented:
- QuTiP already has some interpolation functions that can be compiled down into Cython.  Is there some reason we need new ones?  Does the new solver support the current interp functions?

- I don't see any changes / additions to unit tests.  Are these still coming?

- The new `td_qobj` objects should be able to be passed to all the solvers. This is most easily done by just looking for a `td_qobj` instance and then grabbing the `Q_object` from it.

Ericgig commented:
I removed the malloc and rewrote the examples in a way to discourage the mixing of str and func format.

I have some unit tests for the td_Qobj, but I still have to add new ones for mcsolve.
It seems I do not have the same version of the pickle module than the tests server. Will check that next week.

nonhermitian commented:
Hey @Ericgig, what is the status of this Pull?

Ericgig commented:
The mcsolve is Ok. But I did not test it on windows yet.
td_Qobj work fine for the solvers. However, there are probably a few methods/properties that would be useful but I did not think about. If somebody would have the time to try it and send feedback, that would be great.
For now, sesolve and mesolve also accept td_Qobj. I am working, in another branch, to have the stochastic solvers work with it.

Ericgig commented:
Split in 2 other pull request: mcsolve and td_qobj. Should be removed.
